src/core/lib/json/json_object_loader.h

```diff
@@ -0,0 +1,387 @@
+// Copyright 2020 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_JSON_JSON_OBJECT_LOADER_H
+#define GRPC_CORE_LIB_JSON_JSON_OBJECT_LOADER_H
+
+#include <grpc/support/port_platform.h>
+
+#include <sys/types.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <string>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/json/json.h"
+
+// Provides a means to load JSON objects into C++ objects, with the aim of
+// minimizing object code size.
+//
+// Usage:
+// Given struct Foo:
+//   struct Foo {
+//     int a;
+//     int b;
+//   };
+// We add a member to Foo to declare how to load the object from JSON:
+//   struct Foo {
+//     int a;
+//     int b;
+//     static const JsonLoaderInterface* JsonLoader() {
+//       static const auto loader = JsonObjectLoader<Foo>()
+//           .Field("a", &Foo::a)
```

Comment:
One common use-case we have is when parsing the LB policy config for a non-leaf policy, where the config for the child policy is embedded in the config for the parent policy.  In that case, at the point where we're defining the config for the parent policy, we don't actually know what fields are going to be in the config for the child policy, so we can't define them inline.  Instead, what we want is to pass the `Json` object of the child's config to the LB policy registry, which will parse the config for us.  For example:

https://github.com/grpc/grpc/blob/28db143b98682b06aa5c58059ffe53b42f28b395/src/core/ext/filters/client_channel/lb_policy/priority/priority.cc#L840

For the majority of instances of this use-case, we just want to store the resulting `RefCountedPtr<LoadBalancingPolicy::Config>` object, so I think we can probably just define `AutoLoader<RefCountedPtr<LoadBalancingPolicy::Config>>`.  But we probably don't want to do that here in the JSON library, because we don't want the JSON library to depend on the LB policy API.  Is there a way we can inject a new `AutoLoader<>` type from somewhere in the LB policy library?

Also, there's at least one edge case where we actually want to hang on to the child policy config in `Json` form rather than converting it to `RefCountedPtr<LoadBalancingPolicy::Config>`, since we need to modify it before we actually use it:

https://github.com/grpc/grpc/blob/28db143b98682b06aa5c58059ffe53b42f28b395/src/core/ext/filters/client_channel/lb_policy/rls/rls.cc#L2501

For that kind of case, is there a way that we can pass in a custom parsing function for a particular field, rather than having it be always inferred from the type?  Since the field we want to store here is type `Json`, that seems a little too general to key this solely by the type, so I think we need some sort of escape hatch here.  Or maybe the answer is to define a custom struct type that contains only a `Json` field, and then inject an `AutoLoader<>` implementation for that custom struct type?